core/src/main/scala/caliban/validation/DefaultValue.scala

```diff
@@ -0,0 +1,135 @@
+                s"Failed to parse default value for input '${field.name}': ${e.msg}",
+                "The default value for a field must be written using GraphQL input syntax."
+              )
+            )
+        _     <- Validator.validateInputValues(field, value)
+        _     <- validateInputTypes(field, value)
+      } yield ()
+    }.getOrElse(IO.unit)
+
+  def validateInputTypes(
+    inputValue: __InputValue,
+    argValue: InputValue
+  ): IO[ValidationError, Unit] = validateType(inputValue.`type`(), argValue)
+
+  def validateType(inputType: __Type, argValue: InputValue): IO[ValidationError, Unit] =
+    inputType.kind match {
+      case NON_NULL =>
+        argValue match {
+          case NullValue => failValidation(s"Field is null but was supposed to be NonNull", "")
+          case x         => validateType(inputType.ofType.getOrElse(inputType), x)
+        }
+      case LIST     =>
+        argValue match {
+          case ListValue(values) => IO.foreach_(values)(v => validateType(inputType.ofType.getOrElse(inputType), v))
+          case _                 => failValidation(s"Field has invalid type: $argValue, expected List", "")
+        }
+
+      case INPUT_OBJECT =>
+        argValue match {
+          case ObjectValue(fields) =>
+            IO.foreach_(inputType.inputFields.getOrElse(List.empty)) { f =>
+              val value =
+                fields.collectFirst({ case (name, fieldValue) if name == f.name => fieldValue }).getOrElse(NullValue)
+              validateType(f.`type`(), value)
+            }
+          case _                   => failValidation(s"Field has invalid type: $argValue, expected INPUT_OBJECT", "")
+        }
+      case ENUM         =>
+        argValue match {
+          case EnumValue(value) =>
+            val exists = inputType
+              .enumValues(__DeprecatedArgs(Some(true)))
+              .getOrElse(List.empty)
+              .exists(_.name == value)
+
+            if (exists) IO.unit
+            else failValidation(s"Field has invalid enum value: $value", "")
+          case _                => failValidation(s"Field has invalid type: $argValue", "")
+        }
+      case SCALAR       => validateScalar(inputType, argValue)
+      case x            =>
+        failValidation(s"Field has invalid type $inputType", "")
+    }
+
+  def validateScalar(inputType: __Type, argValue: InputValue) =
+    inputType.name.getOrElse("") match {
+      case "String"  =>
+        argValue match {
+          case StringValue(value) =>
+            IO.unit
+          case t                  => failValidation(s"Field has invalid type $t, expected 'String'", "")
+        }
+      case "ID"      =>
+        argValue match {
+          case StringValue(value) =>
+            IO.unit
+          case t                  => failValidation(s"Field has invalid type $t, expected 'ID'", "")
+        }
+      case "Int"     =>
+        argValue match {
+          case BigIntNumber(value) => IO.unit
```

Comment:
You can simply check if it's a `IntValue` since it's sealed.
Same below with `FloatValue`.

Reply:
👍